### DISCUSSION TITLE
Sort when creating doesn't work with polymrphic pivot models.

### DISCUSSION THREAD

**Discussion by reppair** [freekmurze/eloquent-sortable#134] [General]
Hi there.

I have a many-to-many **polymorphic** relationship with dedicated pivot model, that pivot model extends `Illuminate\Database\Eloquent\Relations\MorphPivot` as described in the [Laravel Docs](https://laravel.com/docs/9.x/eloquent-relationships#defining-custom-intermediate-table-models).
```php
class Skillable extends MorphPivot implements Sortable
{
    use SortableTrait, HasFactory;
    
    ...
```
Thing is, it always sets value of `1` to the sortable column on any new pivot record I create. Same when:
- calling `Skillable::factory()->create()` multiple times
- calling `Skillable::factory()->count(N)->create()`
- calling `Skillable::create($attrs)`

BTW using the Sortable trait and it works as it should on few other regular eloquent modals.

**Comment by freekmurze** [comment]
Feel free to submit a PR that fixes this.

**Reply by reppair** [reply]
Gladly, will give it a try later this week hopefully.

**Comment by swthate** [comment]
Has any progress been made on this? I have a similar pivot table that does not generate any `order` value on create, whether that is via seeding or manually via filament. However, if I alter the order in filament and save it, then order values for all of those relations will be set.

**Reply by reppair** [reply]
Not by me, I've implemented a "workaround" on project level and forgot about it. Anyway, the "workaround" is simple, just have a model observer class for the pivot model (or do it in the model directly) in which you can listen for the `creating` event (or `created`), query the number of records and increment the value of the sorting column. This way any new record will have a properly assigned sort order value based on whatever logic you want.

**Comment by clarkewing** [comment]
For anyone looking for how to make this work appropriately (and future me), the trick is overriding the `buildSortQuery()` method on the pivot table:

```php
public function buildSortQuery(): Builder
{
    return static::query()
        ->where('lesson_id', $this->lesson_id);
}
```
